### PR TITLE
Workarounds for building with Roslyn/MSBuild on Unix.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Allows a successful compile with xplat MSBuild branch and Roslyn on Mono.

Adds the Roslyn package and UseRoslynCompiler property to use Roslyn.
Disable PDB generation (NYI), skip packaging, disables ECMA delay signing.
Also temporarily disables two test projects and VB.
Fixes casing for a few source file includes.

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
@weshaggard, @valmenn, @jaredpar, @ellismg, @stephentoub, @chcosta 

You can build now if you clone and build the MSBuild xplat branch on Ubuntu. MSBuild needs perl, curl, and mono to build (build.pl).

```
mono ~/repos/msbuild/bin/Unix/Debug-MONO/MSBuild.exe -t:Build -fl "-flp:LogFile=msbuild-corefx.log;V=diag;" -p:Configuration=Linux_Debug -p:UseRoslynCompiler=true ~/repos/corefx/build.proj
```

File: dir.props [hunk 1]
```diff
@@ -60,6 +62,41 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(OsEnvironment)'=='Unix' and '$(UseRoslynCompiler)'=='true'">
+    <!--
+      PDB support isn't implemented yet.
+      Note that both DebugSymbols and DebugType need set or project references will assume they need to copy pdbs and fail.
+    -->
+    <DebugSymbols>false</DebugSymbols>
+    <DebugType>none</DebugType>
+    <!-- Delay signing with the ECMA key currently doesn't work. -->
+    <UseECMAKey>false</UseECMAKey>
```
Comment (jaredpar) [inline dir.props]: Wouldn't `<SkipSigning>True</SkipSigning>` be more appropriate here?

File: dir.props [hunk 2]
```diff
@@ -60,6 +62,41 @@
...
+  </PropertyGroup>
+  <ItemGroup Condition="'$(OsEnvironment)'=='Unix' and '$(UseRoslynCompiler)'=='true'">
+    <!-- System.Diagnostics.Debug.Tests currently fails trying to resolve types against System.Runtime -->
+    <ExcludeProjects Include="System.Diagnostics.Debug/tests/System.Diagnostics.Debug.Tests.csproj" />
+    <!-- The OpenSourceSign target complains about System.Diagnostics.FileVersionInfo.TestAssembly.dll not being delay signed. -->
+    <ExcludeProjects Include="System.Diagnostics.FileVersionInfo/tests/System.Diagnostics.FileVersionInfo.Tests.csproj" />
+  </ItemGroup>
+
+  <PropertyGroup Condition="'$(OsEnvironment)'=='Unix'">
+    <!-- Mono currently doesn't include VB targets for portable, notably /lib/mono/xbuild/Microsoft/Portable/v4.5/Microsoft.Portable.VisualBasic.targets. -->
```
Comment (jaredpar) [inline dir.props]: I recently fixed this problem in Mono.  If you build and run the latest from the Mono master branch this will be fixed.  

https://github.com/mono/mono/pull/1726

File: dir.props [hunk 3]
```diff
@@ -151,4 +188,8 @@
     <MakePriExtensionPath>$(_WindowsPhoneKitBinPath)\x86\MrmEnvironmentExtDl.dll</MakePriExtensionPath>
     <MakePriExtensionPath_x64>$(_WindowsPhoneKitBinPath)\x64\MrmEnvironmentExtDl.dll</MakePriExtensionPath_x64>
   </PropertyGroup>
+
+  <Import Project="$(RoslynPropsFile)"
+          Condition="'$(UseRoslynCompiler)'=='true' and Exists('$(RoslynPropsFile)')" />
+
```
Comment (jaredpar) [inline dir.props]: It appears the way this is setup if the user passes `/p:UseRoslynCompiler` but doesn't install the package it will silently revert to the normal csc.  That could give users the false sense they are compiling with Roslyn when in fact they are using the native compiler.   If you remove the `Exists` portion of the `Condition` it would error in that case instead which seems more appropriate.

File: dir.targets [hunk 4]
```diff
@@ -88,11 +88,13 @@
...
     <Copy Condition="Exists('$(NuGetConfigFile)')" SourceFiles="$(NuGetConfigFile)" DestinationFolder="$(IntermediateOutputRootPath)" SkipUnchangedFiles="true" />
 
-    <!-- Add DNU execute rights -->
+    <!-- Add DNU and Roslyn tool execute rights -->
     <Exec Condition="'$(OsEnvironment)'=='Unix'"
           Command="chmod a+x &quot;$(DnxPackageDir)/bin/dnu&quot;" />
     <Exec Condition="'$(OsEnvironment)'=='Unix'"
           Command="chmod a+x &quot;$(DnxPackageDir)/bin/dnx&quot;" />
+    <Exec Condition="'$(OsEnvironment)'=='Unix'"
+          Command="find '$(RoslynPackageDir)tools' -name &quot;*.exe&quot; -exec chmod &quot;+x&quot; '{}' ';'" />
```
Comment (jaredpar) [inline dir.targets]: Curious.  Why was this needed?  In general the compilers are not executed directly but instead by being prefixed with `mono csc.exe`.

File: src/System.IO.Compression/src/System.IO.Compression.csproj [hunk 5]
```diff
@@ -8,7 +8,7 @@
     <OutputType>Library</OutputType>
     <ProjectGuid>{5471BFE8-8071-466F-838E-5ADAA779E742}</ProjectGuid>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <UseECMAKey>true</UseECMAKey>
+    <UseECMAKey Condition="'$(UseECMAKey)'==''">true</UseECMAKey>
```
Comment (akoeplinger) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: I'm curious why this is the only project that needs this?

File: dir.props (hunk 1 above)
Comment (JeremyKuhne) [inline dir.props]: Delay signing with normal keys doesn't fail. In fact, it is necessary to get the right full assembly names. It falls apart hard in tests with InternalsVisibleTo if it's off. Also, open source signing is lumped in with this.

File: dir.props (hunk 2 above)
Comment (JeremyKuhne) [inline dir.props]: Thanks for the info- I'll add in the fix reference.

File: dir.props (hunk 3 above)
Comment (JeremyKuhne) [inline dir.props]: I can't for the initial package restore case. I'll look at adding an initial target that at least warns.

File: dir.targets (hunk 4 above)
Comment (JeremyKuhne) [inline dir.targets]: It failed for me, but I didn't look to hard into it. I copied this logic from @valmenn's change. https://github.com/ValMenn/msbuild/commit/111ee575b1db854459b8228716139a044c6016bc

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: Me too. :) Hopefully someone will chime up or I'll find out sometime today.

**Comment by JeremyKuhne** [comment]
@dotnet-bot please test this

**Comment by akoeplinger** [comment]
@JeremyKuhne This fails on mono 4.0.1 for me due what appears to be a mono bug (), I suppose you tested on 3.12?

```
ResolveAssemblyReferences:
  A TargetFramework profile exclusion list will be generated.

Can't find custom attr constructor image: /home/alexander/dev/corefx/packages/System.Runtime.Extensions/4.0.10-beta-22809/ref/any/System.Runtime.Extensions.dll mtoken: 0x0a000003
* Assertion at class.c:282, condition `!mono_loader_get_last_error ()' not met
```

**Comment by JeremyKuhne** [comment]
@akoeplinger I tested on 3.12.1. In the many builds I ran I did see faults like that, but nothing I could reliably reproduce. I'll see if I can get it to happen again. I assume core dumps are saved somewhere?

```
Mono JIT compiler version 3.12.1 (tarball Fri Mar  6 19:12:47 UTC 2015)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          supported, not enabled.
	GC:            sgen
```

File: dir.props (hunk 1 above)
Comment (jaredpar) [inline dir.props]: Ah okay.  Makes sense.

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: I still don't know *exactly* why this one uses ECMA (looking). I do know that more libraries will be added in the future that also use ECMA.

File: dir.props [hunk 6]
```diff
@@ -151,4 +200,13 @@
...
+  <!-- Building the build.proj should restore successfully, so ignore in that case -->
+  <Target Name="CheckRoslyn" Condition="'$(UseRoslynCompiler)'=='true' and '$(MSBuildProjectFile)'!='build.proj'">
```
Comment (weshaggard) [inline dir.props]: In general I don't like putting targets in props files is there a reason this is here as opposed to in the dir.targets?

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 5 above)
Comment (weshaggard) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: We are trying to preserve the assembly identity of this assembly which has shipped already in .NET and the identity includes the public key which was ECMA in this case.

File: dir.props (hunk 6 above)
Comment (JeremyKuhne) [inline dir.props]: Because it is associated with the Import above it and that does have to come in the props.

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: So was there nothing special beyond preserving the identity? I know this key was treated differently, but I don't fully grok it yet.

**Comment by JeremyKuhne** [comment]
@dotnet-bot please test this

**Comment by JeremyKuhne** [comment]
@dotnet-bot test this please

**Comment by weshaggard** [comment]
@dotnet-bot test this please

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 5 above)
Comment (weshaggard) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: Yes the EMCA key is generally special as it is supposed to be used by libraries that are part of the CLI spec. However it did get used by other assemblies such as this one and so the reason we are using it here is to maintain the same assembly identity.

**Comment by weshaggard** [comment]
As we chatted we should try and move these to the buildtools repo once we test them out in corefx. Otherwise they look good to me and so once we get a clean CI go ahead and merge.

File: dir.targets (hunk 4 above)
Comment (ValMenn) [inline dir.targets]: There are a number of ways tasks can execute external program.  I use mono to run programs that end with .exe, but I only changed Exec task and ToolTask.Execute method. Any task that overrides that method will need to be modified. The alternative is to give execute permission to the .exe file.

**Comment by JeremyKuhne** [comment]
@akoeplinger I've upgraded to 4 now and I see the error you're getting 100% of the time. What can I do to help follow up on this?

**Comment by akoeplinger** [comment]
@JeremyKuhne looks like I narrowed down that crash into a small repro: https://bugzilla.xamarin.com/show_bug.cgi?id=29679

A workaround for that problem is to just do `return null` instead of the code that tries to get the `TargetFrameworkAttribute` (I checked a few of the contract libs and they don't seem to have this set, so shouldn't be found anyway) here in MsBuild: [AssemblyInformation.cs#L252-L280](https://github.com/Microsoft/msbuild/blob/7c87abfd229747f7f6fba02ae55f08f5cd91cbe3/src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs#L252-L280).

While this then allows the build to commence, it hits yet another bug during compilation of the individual assemblies on Mono 4.0.1 (`System.IO.EndOfStreamException: Failed to read past end of stream.` in roslyn), but that one is already fixed on Mono master, so it works now in the end :sparkles:

**Comment by ValMenn** [comment]
@akoeplinger First, I did find this attribute in some assemblies. Previously, it would just throw on an unknown attribute (hence the try/catch -- somewhat unexpected, based on the function description). It looks like it asserts now, which is not caught.

**Comment by JeremyKuhne** [comment]
@akoeplinger I'm having issues with the ref libraries (referenceassemblies-pcl) when I use mono-snapshot. Is there some setup step I need to do besides what is on the main site? http://www.mono-project.com/docs/getting-started/install/linux/ci-packages/

**Comment by akoeplinger** [comment]
@JeremyKuhne the PCL assemblies aren't installed for snapshots. xbuild had an env var to override this, but msbuild doesn't. I just symlinked the stable one from `/usr/lib/mono/xbuild-framworks/.NETPortable` to the snapshot dir in `/opt/mono-snapshot-xxxx/...`.

**Comment by JeremyKuhne** [comment]
@akoeplinger Thanks! What env var was that? Perhaps I can put something similar in MSBuild.

**Comment by akoeplinger** [comment]
@JeremyKuhne `XBUILD_FRAMEWORK_FOLDERS_PATH`, see the [man page](http://manpages.ubuntu.com/manpages/precise/man1/xbuild.1.html#contenttoc4) for more details on how it's implemented.

**Comment by JeremyKuhne** [comment]
@akoeplinger Can you give me a specific setup you were able to get built? Which mono snapshot? Which Ubuntu release? I'm getting a lot of issues, I'd like to try and replicate your working scenario. Thanks!

**Comment by akoeplinger** [comment]
@JeremyKuhne soo, this works on a pristine Ubuntu 12.04 VM (should be fine on 14.04 as well):

``` bash
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
echo "deb http://download.mono-project.com/repo/debian wheezy main" | sudo tee /etc/apt/sources.list.d/mono-xamarin.list
echo "deb http://jenkins.mono-project.com/repo/debian sid main" | sudo tee /etc/apt/sources.list.d/mono-jenkins.list
sudo apt-get update
sudo apt-get install git curl mono-snapshot-20150512101836 referenceassemblies-pcl 

. mono-snapshot mono

sudo ln -s /usr/lib/mono/xbuild-frameworks/.NETPortable/ /opt/mono-20150512101836/lib/mono/xbuild-frameworks/

cd ~/Desktop
git clone https://github.com/Microsoft/MSBuild
git clone https://github.com/dotnet/corefx

cd ~/Desktop/MSBuild/
git checkout xplat

# apply the following PR:
# 
# https://github.com/Microsoft/msbuild/pull/93

perl ./build.pl

cd ~/Desktop/corefx/

# apply the following PRs:
# 
# https://github.com/dotnet/corefx/pull/1713
# https://github.com/dotnet/corefx/pull/1714
# https://github.com/dotnet/corefx/pull/1715

export MONO29679=1
mono ~/Desktop/MSBuild/bin/Unix/Debug-MONO/MSBuild.exe  -t:Build -fl "-flp:LogFile=msbuild-corefx.log;V=diag;" -p:Configuration=Linux_Debug -p:UseRoslynCompiler=true -p:SkipTests=true build.proj 
```

Note that I had to add `-p:SkipTests=true` since there were some issues with the tests (looks like some dnx/targets issue), I'll leave fixing that to you :)
I also once got a segfault during dnu restore (I just tried it again and it worked), I think that is fixed in later dnx versions (which also have a workaround to make restoring on Mono faster), so you should maybe update dnx as well.

**Comment by JeremyKuhne** [comment]
@akoeplinger Thanks :) I'll let you know how it goes for me.

**Comment by JeremyKuhne** [comment]
@akoeplinger Getting a lot of segfaults. Tried moving the DNU package forward, ran into a problem with our build task. Think I've got an answer there. I'll keep you up-to-date and loop you in on any changes I make.
